File: porsh/src/main/java/eu/darken/porter/porsh/PorshConfig.java
package eu.darken.porter.porsh;

import android.annotation.SuppressLint;
import android.os.IBinder;
import android.util.Log;

public class PorshConfig {

    private static final String TAG = "PORSHConfig";

    static final int TRANSACTION_createHost = 0;
    static final int TRANSACTION_setWindowSize = 1;
    static final int TRANSACTION_getExitCode = 2;

    private static IBinder binder;
    private static String interfaceToken;
    private static int transactionCodeStart;
    private static String libraryPath;

    static IBinder getBinder() {
        return binder;
    }

    static String getInterfaceToken() {
        return interfaceToken;
    }

    static int getTransactionCode(int code) {
        return transactionCodeStart + code;
    }

    public static void setLibraryPath(String path) {
        libraryPath = path;
    }

    @SuppressLint("UnsafeDynamicallyLoadedCode")
    private static void loadLibrary() {
        if (libraryPath == null) {
            System.loadLibrary("porsh");
        } else {
            System.load(libraryPath + "/libporsh.so");
        }
    }

    public static void init(String interfaceToken, int transactionCodeStart) {
        Log.d(TAG, "init (server) " + interfaceToken + " " + transactionCodeStart);
        PorshConfig.interfaceToken = interfaceToken;
        PorshConfig.transactionCodeStart = transactionCodeStart;
        loadLibrary();
    }

    public static void init(IBinder binder, String interfaceToken, int transactionCodeStart) {
        Log.d(TAG, "init (client) " + binder + " " + interfaceToken + " " + transactionCodeStart);
        PorshConfig.binder = binder;
        PorshConfig.interfaceToken = interfaceToken;
        PorshConfig.transactionCodeStart = transactionCodeStart;
        loadLibrary();
    }
}

File: porsh/src/main/java/eu/darken/porter/porsh/PorshConfig.kt
package eu.darken.porter.porsh

import android.annotation.SuppressLint
import android.os.IBinder
import android.util.Log

object PorshConfig {

    private const val TAG = "PORSHConfig"

    internal const val TRANSACTION_createHost = 0
    internal const val TRANSACTION_setWindowSize = 1
    internal const val TRANSACTION_getExitCode = 2

    private var binder: IBinder? = null
    private var interfaceToken: String? = null
    private var transactionCodeStart = 0
    private var libraryPath: String? = null

    /** Set by [init]; a client that transacts before that has no wire to speak on. */
    internal fun getBinder(): IBinder = checkNotNull(binder) { "PorshConfig.init was not called" }

    internal fun getInterfaceToken(): String = checkNotNull(interfaceToken) { "PorshConfig.init was not called" }

    internal fun getTransactionCode(code: Int): Int = transactionCodeStart + code

    fun setLibraryPath(path: String?) {
        libraryPath = path
    }

    @SuppressLint("UnsafeDynamicallyLoadedCode")
    private fun loadLibrary() {
        val path = libraryPath
        if (path == null) {
            System.loadLibrary("porsh")
        } else {
            System.load("$path/libporsh.so")
        }
    }

    fun init(interfaceToken: String, transactionCodeStart: Int) {
        Log.d(TAG, "init (server) $interfaceToken $transactionCodeStart")
        this.interfaceToken = interfaceToken
        this.transactionCodeStart = transactionCodeStart
        loadLibrary()
    }

    fun init(binder: IBinder, interfaceToken: String, transactionCodeStart: Int) {
        Log.d(TAG, "init (client) $binder $interfaceToken $transactionCodeStart")
        this.binder = binder
        this.interfaceToken = interfaceToken
        this.transactionCodeStart = transactionCodeStart
        loadLibrary()
    }
}
